File: v1/examples/python/12_websocket_events.py
"""
12_websocket_events.py - TS-API v1 WebSocket Real-time Event Subscription

Subscribes to real-time events (LPR, channelStatus, etc.) via WebSocket.

Two subscription modes:
  1. URL query params:  ?topics=LPR,channelStatus  (subscribe on connect)
  2. Dynamic send():    {"subscribe":"LPR"}         (subscribe after connect, v1 only)

Endpoint:
  ws://host:port/wsapi/v1/events

Auth:
  Header: Authorization: Bearer {accessToken}  (primary)
  Header: X-API-Key: {apiKey}                  (alternative)
  Query:  ?token={accessToken}                 (browser fallback)
  Query:  ?apikey={apiKey}                     (browser fallback)

REQUIRES: pip install websockets
"""

import asyncio
import json
import os
import ssl
import sys

sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
from config import NVR_API_KEY, WS_URL
from http_client import NvrClient

if not NVR_API_KEY:
    print('NVR_API_KEY environment variable is required')
    sys.exit(1)

try:
    import websockets
except ImportError:
    print('REQUIRES: pip install websockets')
    sys.exit(1)


async def main():
    client = NvrClient()
    client.set_api_key(NVR_API_KEY)

    ssl_ctx = ssl.SSLContext(ssl.PROTOCOL_TLS_CLIENT)
    ssl_ctx.check_hostname = False
    ssl_ctx.verify_mode = ssl.CERT_NONE
    headers = {'X-API-Key': NVR_API_KEY}

    # ── Method 1: Subscribe via URL query params (classic) ──
    print('=== Method 1: Subscribe via URL (10 seconds) ===')

    ws_url = f'{WS_URL}/wsapi/v1/events?topics=LPR,channelStatus&apikey={NVR_API_KEY}'
    messages = []

    try:
        async with websockets.connect(ws_url, additional_headers=headers, ssl=ssl_ctx) as ws:
            print('  Connected!')

            async def receive_events():
                try:
                    async for data in ws:
                        try:
                            msg = json.loads(data)
                            messages.append(msg)
                            print(f'  [{msg.get("topic", msg.get("type", "?"))}] {json.dumps(msg)}')
                        except json.JSONDecodeError:
                            print(f'  Raw: {data}')
                except websockets.ConnectionClosed:
                    pass

            try:
                await asyncio.wait_for(receive_events(), timeout=10.0)
            except asyncio.TimeoutError:
                pass

            print(f'  Received {len(messages)} events')

    except Exception as e:
        print(f'  Failed: {e}')

    # ── Method 2: Dynamic subscribe/unsubscribe via send() (v1 only) ──
    #   - Connect WITHOUT topics
    #   - Subscribe/unsubscribe at any time
    #   - Per-topic filters (ch, objectTypes, lot, spot)
    #   - Re-subscribe to update filters
    print('\n=== Method 2: Dynamic Subscribe (10 seconds) ===')

    ws_url2 = f'{WS_URL}/wsapi/v1/events'
    messages2 = []

    try:
        async with websockets.connect(ws_url2, additional_headers=headers, ssl=ssl_ctx) as ws:
            print('  Connected (no topics yet)')

            # Phase 1: Subscribe to initial topics with per-topic filters
            print('  [Phase 1] Subscribe channelStatus + LPR (ch 1,2)')
            await ws.send(json.dumps({'subscribe': 'channelStatus'}))
            await ws.send(json.dumps({'subscribe': 'LPR', 'ch': [1, 2]}))

            async def receive_dynamic():
                try:
                    async for data in ws:
                        try:
                            msg = json.loads(data)
                            messages2.append(msg)

                            # Handle control responses
                            if msg.get('type') == 'subscribed':
                                print(f'  Subscribed to: {msg.get("topic")}')
                                continue
                            if msg.get('type') == 'unsubscribed':
                                print(f'  Unsubscribed from: {msg.get("topic")}')
                                continue
                            if msg.get('type') == 'error':
                                print(f'  Error: {msg.get("message")} (topic: {msg.get("topic", "N/A")})')
                                continue

                            # Handle event data
                            print(f'  [{msg.get("topic", "?")}] {json.dumps(msg)}')
                        except json.JSONDecodeError:
                            print(f'  Raw: {data}')
                except websockets.ConnectionClosed:
                    pass

            async def dynamic_flow():
                recv_task = asyncio.create_task(receive_dynamic())

                # Phase 2 (3s): Add new topic + update existing filter
                await asyncio.sleep(3)
                print('  [Phase 2] Add object topic + expand LPR to ch 1-4')
                await ws.send(json.dumps({'subscribe': 'object', 'objectTypes': ['human', 'vehicle']}))
                await ws.send(json.dumps({'subscribe': 'LPR', 'ch': [1, 2, 3, 4]}))

                # Phase 3 (6s): Unsubscribe topic + subscribe new topic + reduce channels
                await asyncio.sleep(3)
                print('  [Phase 3] Unsubscribe channelStatus + add motionChanges (ch 1) + reduce LPR to ch 1,3')
                await ws.send(json.dumps({'unsubscribe': 'channelStatus'}))
                await ws.send(json.dumps({'subscribe': 'motionChanges', 'ch': [1]}))
                await ws.send(json.dumps({'subscribe': 'LPR', 'ch': [1, 3]}))  # re-subscribe with fewer ch drops ch 2,4

                await asyncio.sleep(4)
                recv_task.cancel()

            try:
                await asyncio.wait_for(dynamic_flow(), timeout=11.0)
            except (asyncio.TimeoutError, asyncio.CancelledError):
                pass

            print(f'  Received {len(messages2)} messages')

    except Exception as e:
        print(f'  Failed: {e}')


if __name__ == '__main__':
    asyncio.run(main())

File: v1/examples/swift/12-websocket-events.swift
/**
 * Example 12: WebSocket - Real-time Event Subscription
 *
 * Demonstrates subscribing to real-time events via WebSocket.
 * Topics: LPR, channelStatus, emergencyCall, object, recording
 *
 * Two subscription modes:
 *   1. URL query params:  ?topics=LPR,channelStatus  (subscribe on connect)
 *   2. Dynamic send():    {"subscribe":"LPR"}         (subscribe after connect, v1 only)
 *
 * Endpoint:
 *   ws://host:port/wsapi/v1/events
 *
 * Auth:
 *   Header: X-API-Key: {apiKey}                  (primary)
 *   Header: Authorization: Bearer {accessToken}   (alternative)
 *   Query:  ?apikey={apiKey}                      (browser fallback)
 *   Query:  ?token={accessToken}                  (browser fallback)
 *
 * Build: swiftc -o 12-websocket-events NvrClient.swift 12-websocket-events.swift
 * Run:   NVR_HOST=192.168.0.100 NVR_API_KEY=tsapi_key_... ./12-websocket-events
 */

import Foundation

#if canImport(FoundationNetworking)
import FoundationNetworking
#endif

@main
struct App {
  static func main() async throws {
    let client = NvrClient()
    guard !client.apiKey.isEmpty else {
        print("NVR_API_KEY environment variable is required")
        exit(1)
    }
    client.setApiKey(client.apiKey)

    // -------------------------------------------------
    // Method 1: Subscribe via URL query params (classic)
    // -------------------------------------------------
    print("=== Method 1: Subscribe via URL (10 seconds) ===")

    let wsTask1 = client.websocket(
        path: "/wsapi/v1/events?topics=LPR,channelStatus"
    )

    wsTask1.resume()
    print("  Connected!")

    var messageCount1 = 0
    let deadline1 = Date().addingTimeInterval(10.0)
    var wsError1: Error? = nil

    while Date() < deadline1 {
        do {
            let message = try await withThrowingTaskGroup(of: URLSessionWebSocketTask.Message?.self) { group in
                group.addTask {
                    try await wsTask1.receive()
                }
                group.addTask {
                    let remaining = deadline1.timeIntervalSinceNow
                    if remaining > 0 {
                        try await Task.sleep(nanoseconds: UInt64(remaining * 1_000_000_000))
                    }
                    return nil
                }

                let result = try await group.next()
                group.cancelAll()
                return result ?? nil
            }

            guard let message = message else { break }

            switch message {
            case .string(let text):
                messageCount1 += 1
                if let data = text.data(using: .utf8),
                   let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] {
                    let topic = json["topic"] ?? json["type"] ?? "?"
                    print("  [\(topic)] \(text)")
                } else {
                    print("  Raw: \(text)")
                }
            case .data(let data):
                messageCount1 += 1
                print("  Binary: \(data.count) bytes")
            @unknown default:
                break
            }
        } catch {
            wsError1 = error
            break
        }
    }

    if let err = wsError1 {
        let desc = "\(err)"
        if desc.contains("SEC_E_") || desc.contains("certificate") || desc.contains("SSL") {
            print("  WebSocket SSL error (self-signed cert not supported on this platform)")
            print("  Use NVR_SCHEME=http for WebSocket examples on Windows")
        } else {
            print("  WebSocket error: \(err)")
        }
    }
    print("  Received \(messageCount1) events")
    wsTask1.cancel(with: .normalClosure, reason: nil)
    try await Task.sleep(nanoseconds: 500_000_000)

    // -------------------------------------------------
    // Method 2: Dynamic subscribe/unsubscribe via send() (v1 only)
    //   - Connect WITHOUT topics
    //   - Subscribe/unsubscribe at any time
    //   - Per-topic filters (ch, objectTypes, lot, spot)
    //   - Re-subscribe to update filters
    // -------------------------------------------------
    print("\n=== Method 2: Dynamic Subscribe (10 seconds) ===")

    let wsTask2 = client.websocket(
        path: "/wsapi/v1/events"
    )

    wsTask2.resume()
    print("  Connected (no topics yet)")

    // Dynamically subscribe to topics with per-topic filters
    do {
        try await wsTask2.send(.string("{\"subscribe\":\"channelStatus\"}"))
        try await wsTask2.send(.string("{\"subscribe\":\"LPR\",\"ch\":[1,2]}"))
        try await wsTask2.send(.string("{\"subscribe\":\"object\",\"objectTypes\":[\"human\",\"vehicle\"]}"))
    } catch {
        let desc = "\(error)"
        if desc.contains("SEC_E_") || desc.contains("certificate") || desc.contains("SSL") {
            print("  WebSocket SSL error (self-signed cert not supported on this platform)")
            print("  Use NVR_SCHEME=http for WebSocket examples on Windows")
        } else {
            print("  WebSocket send error: \(error)")
        }
        wsTask2.cancel(with: .normalClosure, reason: nil)
        try await Task.sleep(nanoseconds: 500_000_000)
        return
    }

    var messageCount2 = 0
    let deadline2 = Date().addingTimeInterval(10.0)
    var didUpdateFilters = false

    while Date() < deadline2 {
        // After 3 seconds: unsubscribe channelStatus, update LPR filter
        if !didUpdateFilters && Date().addingTimeInterval(-3.0) > deadline2.addingTimeInterval(-10.0) {
            didUpdateFilters = true
            print("  --- Unsubscribing channelStatus, updating LPR filter ---")
            try? await wsTask2.send(.string("{\"unsubscribe\":\"channelStatus\"}"))
            try? await wsTask2.send(.string("{\"subscribe\":\"LPR\",\"ch\":[1,2,3,4]}"))
        }

        do {
            let message = try await withThrowingTaskGroup(of: URLSessionWebSocketTask.Message?.self) { group in
                group.addTask {
                    try await wsTask2.receive()
                }
                group.addTask {
                    let remaining = deadline2.timeIntervalSinceNow
                    if remaining > 0 {
                        try await Task.sleep(nanoseconds: UInt64(remaining * 1_000_000_000))
                    }
                    return nil
                }

                let result = try await group.next()
                group.cancelAll()
                return result ?? nil
            }

            guard let message = message else { break }

            switch message {
            case .string(let text):
                messageCount2 += 1
                if let data = text.data(using: .utf8),
                   let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] {
                    let type = json["type"] as? String ?? ""
                    let topic = json["topic"] as? String ?? ""

                    // Handle control responses
                    switch type {
                    case "subscribed":
                        print("  Subscribed to: \(topic)")
                    case "unsubscribed":
                        print("  Unsubscribed from: \(topic)")
                    case "error":
                        let msg = json["message"] as? String ?? ""
                        print("  Error: \(msg) (topic: \(topic))")
                    default:
                        // Event data
                        let t = topic.isEmpty ? "?" : topic
                        print("  [\(t)] \(text)")
                    }
                } else {
                    print("  Raw: \(text)")
                }
            case .data(let data):
                messageCount2 += 1
                print("  Binary: \(data.count) bytes")
            @unknown default:
                break
            }
        } catch {
            break
        }
    }

    print("  Received \(messageCount2) messages")
    wsTask2.cancel(with: .normalClosure, reason: nil)

    try await Task.sleep(nanoseconds: 500_000_000)
  }
}

/*
 * ─────────────────────────────────────────────────
 * LPR Event Compatibility
 * ─────────────────────────────────────────────────
 *
 * LPR events may arrive in two formats:
 *
 *   v1.0.0 (single plate):  { "topic": "LPR", "plateNo": "12가3456", ... }
 *   v1.0.1 (batch/array):   { "topic": "LPR", "plates": [ { "plateNo": "12가3456", ... }, ... ] }
 *
 * To handle both formats transparently:
 *
 *   if let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
 *      json["topic"] as? String == "LPR" {
 *       let plates: [[String: Any]]
 *       if let arr = json["plates"] as? [[String: Any]] {
 *           plates = arr           // v1.0.1 batch format
 *       } else {
 *           plates = [json]        // v1.0.0 single-plate format
 *       }
 *       for p in plates {
 *           let plateNo = p["plateNo"] as? String ?? ""
 *           let score = p["score"] as? Double ?? 0
 *           print("Plate: \(plateNo)  Score: \(score)")
 *       }
 *   }
 */
